src/Dsll.hs

```diff
@@ -203,90 +211,111 @@ pd1 = do
+                                  else error "Error!" -- Does not suppose to reach here
+
+checkDsllStmt e (OdStmt v) = let kinds = seconds (varsOd v)
+                                 env1 = foldl checkK e kinds
+                                 localEnv = foldl updateEnv env1 (varsOd v)
+                                 args = seconds (typesOd v)
+                                 res = toOd v
+                                 env2 = foldl checkT localEnv args
+                                 temp = checkT env2 res
+                                 op = Operator { nameop = nameOd v, ylsop = firsts (varsOd v),
+                                                 kindsop = seconds (varsOd v), tlsop = seconds (typesOd v), top = toOd v }
+                                 ef = addName (nameOd v) e
+                               in if env2 == e || env2 /= e && temp == e || temp /= e
+                                   then ef { operators = M.insert (nameOd v) op $ operators ef }
+                                   else error "Error!"  -- Does not suppose to reach here
+
+
+checkDsllStmt e (PdStmt (Pd1Const v)) = let kinds = seconds (varsPd1 v)
+                                            env1 = foldl checkK e kinds
+                                            localEnv = foldl updateEnv env1 (varsPd1 v)
+                                            args = seconds (typesPd1 v)
+                                            env2 = foldl checkT localEnv args
+                                            pd1 = Pred1 $ Prd1 { namepred1 = namePd1 v,
+                                                                 ylspred1  = firsts (varsPd1 v),
+                                                                 kindspred1  = seconds (varsPd1 v),
+                                                                 tlspred1  = seconds (typesPd1 v),
+                                                                 ppred1    = toPd1 v }
+                                            ef = addName (namePd1 v) e
+                                        in if env2 == e || env2 /= e
+                                            then ef { predicates = M.insert (namePd1 v) pd1 $ predicates ef }
+                                            else error "Error!"  -- Does not suppose to reach here
+
+checkDsllStmt e (PdStmt (Pd2Const v)) = let pd = Pred2 $ Prd2 { namepred2 = namePd2 v, plspred2 = seconds (propsPd2 v),
+                                                                ppred2 = toPd2 v }
+                                            ef = addName (namePd2 v) e
+                                         in ef { predicates = M.insert (namePd2 v) pd $ predicates ef }
+
+computeSubTypes :: VarEnv -> VarEnv
+computeSubTypes e = let env1 = e { subTypes = transitiveClosure (subTypes e)}
+                    in if isClosureNotCyclic (subTypes env1) then
+                      env1
+                    else env1 { errors = errors env1 ++ "Cyclic Subtyping Relation! \n"}
```

Comment:
after storing a list of errors, you can remove the newlines from error strs

Reply:
I keep them since I have one big string of all the errors, seperated by newlines

Reply:
I'd prefer if the `VarEnv` stored a list of strings. That would be cleaner since you can choose later how you want to display the errors